File: code/c/12-merge_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

/**
 * Function to merge two subarrays of arr[].
 * First subarray is arr[left..mid]
 * Second subarray is arr[mid+1..right]
 *
 * @param arr The main array containing both subarrays.
 * @param left The starting index of the first subarray.
 * @param mid The ending index of the first subarray.
 * @param right The ending index of the second subarray.
 */
void merge(int arr[], int left, int mid, int right) {
  int n1 = mid - left + 1;
  int n2 = right - mid;

  // Create temp arrays
  int *L = (int *)malloc(n1 * sizeof(int));
  int *R = (int *)malloc(n2 * sizeof(int));

  // Copy data to temp arrays L[] and R[]
  for (int i = 0; i < n1; i++) L[i] = arr[left + i];
  for (int j = 0; j < n2; j++) R[j] = arr[mid + 1 + j];

  // Merge the temp arrays back into arr[left..right]
  int i = 0;     // Initial index of first subarray
  int j = 0;     // Initial index of second subarray
  int k = left;  // Initial index of merged subarray

  while (i < n1 && j < n2) {
    if (L[i] <= R[j]) {
      arr[k++] = L[i++];
    } else {
      arr[k++] = R[j++];
    }
  }

  // Copy the remaining elements of L[], if any
  while (i < n1) arr[k++] = L[i++];

  // Copy the remaining elements of R[], if any
  while (j < n2) arr[k++] = R[j++];

  // Free the temporary arrays
  free(L);
  free(R);
}

/**
 * Function to perform Merge Sort on an array.
 *
 * @param arr The array to sort.
 * @param left The starting index of the array.
 * @param right The ending index of the array.
 */
void merge_sort(int arr[], int left, int right) {
  if (left < right) {
    // Same as (left + right)/2 but avoids overflow
    int mid = left + (right - left) / 2;

    // Sort first and second halves
    merge_sort(arr, left, mid);
    merge_sort(arr, mid + 1, right);

    merge(arr, left, mid, right);
  }
}

/**
 * Main function to demonstrate Merge Sort and measure execution time.
 */
int main() {
  int N_values[] = {100000, 200000, 400000, 800000};
  int num_tests = sizeof(N_values) / sizeof(N_values[0]);

  for (int t = 0; t < num_tests; t++) {
    int N = N_values[t];
    int *arr = (int *)malloc(N * sizeof(int));

    // Seed the random number generator
    srand(time(NULL));

    // Fill the array with random integers
    for (int i = 0; i < N; i++) {
      arr[i] = rand();
    }

    // Measure the start time
    clock_t start_time = clock();

    // Perform Merge Sort
    merge_sort(arr, 0, N - 1);

    // Measure the end time
    clock_t end_time = clock();

    // Calculate the elapsed time in seconds
    double time_spent = (double)(end_time - start_time) / CLOCKS_PER_SEC;

    // Verify that the array is sorted
    int is_sorted = 1;
    for (int i = 1; i < N; i++) {
      if (arr[i - 1] > arr[i]) {
        is_sorted = 0;
        break;
      }
    }

    // Print the result and time
    printf("Merge Sort with N = %d\n", N);
    if (is_sorted)
      printf("Array is sorted.\n");
    else
      printf("Array is NOT sorted.\n");
    printf("Time taken: %f seconds\n\n", time_spent);

    // Free the allocated memory
    free(arr);
  }

  return 0;
}

File: code/zig/12-merge_sort.zig
const std = @import("std");
const print = std.debug.print;
const time = std.time;
const Instant = time.Instant;
const allocator = std.heap.page_allocator;
const Allocator = std.mem.Allocator;
const rand = std.rand;

/// Function to perform Merge Sort on an array.
///
/// # Parameters
///
/// - `allocator`: The allocator to use for temporary arrays.
/// - `arr`: The array to sort.
/// - `left`: The starting index of the array.
/// - `right`: The ending index of the array.
fn mergeSort(alloc: Allocator, arr: []usize, left: usize, right: usize) !void {
    if (left < right) {
        // Same as (left + right)/2 but avoids overflow
        const mid: usize = left + @divFloor(right - left, 2);

        // Sort first and second halves
        try mergeSort(alloc, arr, left, mid);
        try mergeSort(alloc, arr, mid + 1, right);

        try merge(alloc, arr, left, mid, right);
    }
}

/// Function to merge two subarrays of `arr`.
/// First subarray is `arr[left..mid+1]`
/// Second subarray is `arr[mid+1..right+1]`
///
/// # Parameters
///
/// - `allocator`: The allocator to use for temporary arrays.
/// - `arr`: The main array containing both subarrays.
/// - `left`: The starting index of the first subarray.
/// - `mid`: The ending index of the first subarray.
/// - `right`: The ending index of the second subarray.
fn merge(alloc: Allocator, arr: []usize, left: usize, mid: usize, right: usize) !void {
    const n1 = mid - left + 1;
    const n2 = right - mid;

    // Create temp arrays
    var L = try alloc.alloc(usize, n1);
    defer alloc.free(L);
    var R = try alloc.alloc(usize, n2);
    defer alloc.free(R);

    // Copy data to temp arrays L[] and R[]
    for (0..n1) |i| L[i] = arr[left + i];
    for (0..n2) |j| R[j] = arr[mid + 1 + j];

    // Merge the temp arrays back into arr[left..right]
    var i: usize = 0; // Initial index of first subarray
    var j: usize = 0; // Initial index of second subarray
    var k: usize = left; // Initial index of merged subarray

    while (i < n1 and j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i += 1;
        } else {
            arr[k] = R[j];
            j += 1;
        }
        k += 1;
    }

    // Copy the remaining elements of L[], if any
    while (i < n1) : ({
        i += 1;
        k += 1;
    }) {
        arr[k] = L[i];
    }

    // Copy the remaining elements of R[], if any
    while (j < n2) : ({
        j += 1;
        k += 1;
    }) {
        arr[k] = R[j];
    }
}

/// Function to verify if an array is sorted.
///
/// # Parameters
///
/// - `arr`: The array to check.
fn isSorted(arr: []usize) bool {
    const n = arr.len;
    for (0..n - 1) |i| {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

/// Main function to demonstrate Merge Sort and measure execution time.
pub fn main() !void {
    const N_values = [_]usize{ 10_000, 20_000, 30_000, 40_000 };

    // Seed the random number generator
    const seed: u64 = @intCast(time.nanoTimestamp());
    var rng = rand.DefaultPrng.init(seed); // Xoshiro256 is good enough

    for (N_values) |N| {
        // Allocate memory for the array
        var arr = try allocator.alloc(usize, N);
        defer allocator.free(arr);

        // Fill the array with random integers
        for (0..N) |i| {
            const random_value: usize = rng.next() % N;
            arr[i] = random_value;
        }

        // Measure the start time
        const start = try Instant.now();

        // Perform Merge Sort
        const left: usize = 0;
        const right = arr.len - 1;
        try mergeSort(allocator, arr, left, right);

        // Measure the end time
        const end = try Instant.now();

        // Calculate the elapsed time in seconds
        const elapsed: f64 = @floatFromInt(end.since(start));
        const time_spent = elapsed / time.ns_per_s;

        // Verify that the array is sorted
        const is_sorted = isSorted(arr);

        // Print the result and time
        print("Merge Sort with N = {d}\n", .{N});
        if (is_sorted) {
            print("Array is sorted.\n", .{});
        } else {
            print("Array is NOT sorted.\n", .{});
        }
        print("Time taken: {d} seconds\n\n", .{time_spent});
    }
}
